hadoop-hdds/docs/content/design/event-notifications.md

```diff
@@ -0,0 +1,193 @@
+This is similar to https://issues.apache.org/jira/browse/HDDS-5984 but aims to encapsulate all events and not solely S3 buckets.  
+This document proposes a potential solution and discusses some of the challenges/open questions.
+
+# Introduction
+
+Apache Ozone does not currently provide the ability to consume filesystem events, similar to how HDFS does with Inotify or S3 with bucket notifications.  
+These events are an integral part of integration with external systems to support real-time, scalable, and programmatic monitoring of changes in the data or metadata stored in Ozone.  
+These external systems can use notifications of objects created/deleted to trigger data processing workflows, replication and monitoring alerts.
+
+# Goals
+
+Provide support for all events across the Ozone filesystem for FSO and non FSO buckets, including renames and changes to acls.
+Not impact with performance of client requests.
+Guarantee at-least-once delivery.
+
+# Non-Goals
+
+Filtering of events or paths/buckets
+Persistent storage of notification messages
+Asynchronous delivery
+
+# Supported OMRequests
+
+OMDirectoryCreateRequest
+OMKeyCommitRequest
+OMKeyDeleteRequest
+OMKeyRenameRequest
+OMKeyAddAclRequest
+OMKeyRemoveAclRequest
+OMKeySetAclRequest
+OMKeySetTimesRequest
+
+# Design
+
+## Overview
+
+Introduce a new OzoneManager node in the ratis ring in LISTENER mode.  
+The node will be defined up-front in ozone-site.xml.
+This node will maintain metadata similar to the other OM nodes however it will generate notifications for events after they have been successfully commited.
+This node will not be able to become a leader, so the notification process will not impact client requests.  
+
+![OzoneEventNotification.png](OzoneEventNotification.png)
+
+#### Component
+
+Implementation of this feature requires changes to the OzoneManager:
+    - Implement an agnostic notification interface
+    - Add support for a LISTENER node
+    - Add a hook in the OMRequest exection flow to generate notifications
+
+#### Component
+
+Introduces a new field on OMAdminProtocol.proto to identify the notify node:
+
+```protobuf
+message OMNodeInfo {
+required string nodeID = 1;
+required string hostAddress = 2;
+required uint32 rpcPort = 3;
+required uint32 ratisPort = 4;
+optional NodeState nodeState = 5 [default=ACTIVE];
+optional bool isNotifyNode = 6 [default=false];
+}
+```
+
+## Performance
+
+While this is a synchronous only approach any latency between notification target and OM should not impact the performance of client requests as the notification does not run on the leader.
+
+## Configuration
+
+This is the configuration to enable event notification on node on1.
+
+```xml
+<property>
+  <name>ozone.event.notification.node</name>
+  <value>on1</value>
+</property>
+```
+Example configuration to provide a kafka instance:
+```xml
+
+<configration>
```

Comment:
Maybe as a future work, make it possible to configure notification on a per bucket basis, which is actually more AWS-like. That way each bucket owner can decide to send notification or send them to different kafka topics, for example.

Reply:
Optionally, support AWS's PutBucketNotificationConfiguration and GetBucketNotificationConfiguration